cpp/src/arrow/util/trie_test.cc

```diff
@@ -175,6 +175,15 @@ TEST(Trie, EmptyString) {
   ASSERT_EQ(-1, trie.Find("x"));
 }
 
+TEST(Trie, LongString) {
+  TrieBuilder builder;
+  ASSERT_OK(builder.Append(""));
+  const Trie trie = builder.Finish();
+  auto maxlen = static_cast<size_t>(std::numeric_limits<int16_t>::max()) + 1;
```

Comment:
We may want to test with a bunch of limit-case sizes (e.g. `2^n, 2^n-1, 2^n+1` for `n` in 15, 16, 17).

Reply:
I did this and split the cases into good cases and "good & bad" cases.  For the good cases I confirmed we can insert and match.  For the "good & bad" cases I confirmed we don't match the empty string.  I tested on Mac and the test failed (it matched the empty string).  I then added your patch and it passed (well, it failed because of the >/>= issue but passed after that).

I'm testing 2^15-2, 2^15-1, 2^14 (good) and 2^15, 2^16-2 (bad).  Do you think that covers it?

Reply:
Yes, thank you.